### PR TITLE
-ast-dump-decl-types crashes on codes involving concepts #94928

### LINKED ISSUE #94928: `-ast-dump-decl-types` crashes on codes involving concepts
<details><summary>backtrace</summary>
<p>

```
clang++: /root/llvm-project/llvm/include/llvm/Support/Casting.h:662: decltype(auto) llvm::dyn_cast(From*) [with To = clang::ValueDecl; From = clang::Decl]: Assertion `detail::isPresent(Val) && "dyn_cast on a non-existent value"' failed.

 #0 0x0000000003a01968 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3a01968)
 #1 0x00000000039ff64c llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x39ff64c)
 #2 0x0000000003948068 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007a2c9c642520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007a2c9c6969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x00007a2c9c642476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007a2c9c6287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00007a2c9c62871b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #8 0x00007a2c9c639e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
 #9 0x00000000045cb5a5 (anonymous namespace)::ASTPrinter::print(clang::Decl*) ASTConsumers.cpp:0:0
#10 0x00000000045cd6ef (anonymous namespace)::ASTPrinter::TraverseDecl(clang::Decl*) ASTConsumers.cpp:0:0
#11 0x00000000045cda6a clang::RecursiveASTVisitor<(anonymous namespace)::ASTPrinter>::TraverseDeclContextHelper(clang::DeclContext*) (.part.0) ASTConsumers.cpp:0:0
#12 0x00000000045d72f9 clang::RecursiveASTVisitor<(anonymous namespace)::ASTPrinter>::TraverseTranslationUnitDecl(clang::TranslationUnitDecl*) ASTConsumers.cpp:0:0
#13 0x00000000045cd670 (anonymous namespace)::ASTPrinter::TraverseDecl(clang::Decl*) ASTConsumers.cpp:0:0
#14 0x000000000635b0ac clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x635b0ac)
...
```

</p>
</details> 

Code:
```cpp
namespace ns {

template <typename T>
concept C = true;

} // namespace ns

using ns::C;

template <typename T> concept Foo = C<T>;

template <C... T, C auto U>
auto FooFunc(C auto V) -> C decltype(auto) {
  C auto W = V;
  return W;
}
```

Command line:
```
-std=c++20 -Xclang -ast-dump -Xclang -ast-dump-decl-types -Xclang -ast-dump-filter="Foo"
```

https://godbolt.org/z/ebcMq8b4e

### PR DESCRIPTION
**Pull request by ofAlpaca**
Resolve #94928

Hi @zyn0217
This PR adds `if (TD->getTemplateDecl())` to prevent `InnerD` becoming `nullptr`, suggested by @firstmoonlight.
I also add `-ast-dump-decl-types` option and declare type `CHECK` to the testcase `clang/test/AST/ast-dump-concepts.cpp`.
Could you kindly review this commit?

### CODE REVIEW

File: clang/lib/Frontend/ASTConsumers.cpp [hunk 1]
```diff
@@ -101,7 +101,8 @@ namespace {
       if (DumpDeclTypes) {
         Decl *InnerD = D;
         if (auto *TD = dyn_cast<TemplateDecl>(D))
-          InnerD = TD->getTemplatedDecl();
+          if (TD->getTemplatedDecl())
+            InnerD = TD->getTemplatedDecl();
```
Comment (AaronBallman) [inline clang/lib/Frontend/ASTConsumers.cpp:105]:
```suggestion
          if (Decl *TempD = TD->getTemplatedDecl())
            InnerD = TempD;
```

**Review by AaronBallman — COMMENTED** [review]
Thank you for the fix! Can you also add a release note to `clang/docs/ReleaseNotes.rst` so that users know about the fix?

Also, please add the example code from the linked issue to `ast-dump-concepts.cpp` so that we more clearly demonstrate that the issue was fixed.

**Review by zyn0217 — APPROVED** [review]
LGTM apart from Aaron’s comments. Thanks for the patch!

**Comment by ofAlpaca** [comment]
> Also, please add the example code from the linked issue to `ast-dump-concepts.cpp` so that we more clearly demonstrate that the issue was fixed.

Hi @AaronBallman 
The example code provided by @zyn0217 has already in the `ast-dump-concepts.cpp`.

**Comment by AaronBallman** [comment]
> > Also, please add the example code from the linked issue to `ast-dump-concepts.cpp` so that we more clearly demonstrate that the issue was fixed.
> 
> Hi @AaronBallman The example code provided by @zyn0217 has already in the `ast-dump-concepts.cpp`.

Ah, I missed it because the code was split up in the file. Thanks!

**Review by AaronBallman — APPROVED** [review]
LGTM

**Pull request merged**